pkg/nsx/services/subnetport/store.go

```diff
@@ -0,0 +1,109 @@
+package subnetport
+
+import (
+	"errors"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+)
+
+// keyFunc is used to get the key of a resource, usually, which is the ID of the resource
+func keyFunc(obj interface{}) (string, error) {
+	switch v := obj.(type) {
+	case model.SegmentPort:
+		return *v.Id, nil
+	case types.UID:
+		return string(v), nil
+	default:
+		return "", errors.New("keyFunc doesn't support unknown type")
+	}
+}
+
+// indexFunc is used to get index of a resource, usually, which is the UID of the CR controller reconciles,
+// index is used to filter out resources which are related to the CR
+func indexFunc(obj interface{}) ([]string, error) {
+	res := make([]string, 0, 5)
+	switch o := obj.(type) {
+	case model.SegmentPort:
+		return filterTag(o.Tags), nil
+	default:
+		return res, errors.New("indexFunc doesn't support unknown type")
+	}
+}
+
+var filterTag = func(v []model.Tag) []string {
+	res := make([]string, 0, 5)
+	for _, tag := range v {
+		if *tag.Scope == common.TagScopeSubnetPortCRUID {
+			res = append(res, *tag.Tag)
+		}
+	}
+	return res
+}
+
+// SubnetPortStore is a store for SubnetPorts
+type SubnetPortStore struct {
+	common.ResourceStore
+}
+
+func (vs *SubnetPortStore) Operate(i interface{}) error {
+	if i == nil {
+		return nil
+	}
+	subnetPort := i.(*model.SegmentPort)
+	if subnetPort.MarkedForDelete != nil && *subnetPort.MarkedForDelete {
+		err := vs.Delete(*subnetPort)
+		log.V(1).Info("delete SubnetPort from store", "subnetport", subnetPort)
+		if err != nil {
+			return err
+		}
+	} else {
+		err := vs.Add(*subnetPort)
+		log.V(1).Info("add SubnetPort to store", "subnetport", subnetPort)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (subnetPortStore *SubnetPortStore) GetByKey(key string) *model.SegmentPort {
+	var subnetPort model.SegmentPort
+	obj := subnetPortStore.ResourceStore.GetByKey(key)
+	if obj != nil {
+		subnetPort = obj.(model.SegmentPort)
+	}
+	return &subnetPort
+}
+
+func (subnetPortStore *SubnetPortStore) GetByIndex(key string, value string) []model.SegmentPort {
+	securityPolicies := make([]model.SegmentPort, 0)
+	objs := subnetPortStore.ResourceStore.GetByIndex(key, value)
+	for _, subnetPort := range objs {
+		securityPolicies = append(securityPolicies, subnetPort.(model.SegmentPort))
+	}
+	return securityPolicies
+}
+
+func (vs *SubnetPortStore) GetSubnetPortsByNamespace(ns string) []model.SegmentPort {
+	var ret []model.SegmentPort
+	subnetPorts := vs.List()
+	if len(subnetPorts) == 0 {
+		log.V(1).Info("No subnet port found in SubnetPort store")
+		return ret
+	}
+
+	for _, subnetPort := range subnetPorts {
+		msubnetport := subnetPort.(model.SegmentPort)
+		tags := msubnetport.Tags
+		for _, tag := range tags {
+			if *tag.Scope == common.TagScopeNamespace && *tag.Tag == ns {
```

Comment:
How about creating index for `common.TagScopeNamespace` like `common.TagScopeSubnetPortCRUID`, and leverage functions like `getByIndex` to perform searches?

Reply:
Thanks for the suggestion, added a TODO for it.